Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`

mikaylagawarecki opened the pull request:
Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with @cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.


Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85593

Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

mikaylagawarecki committed:
Add python  and  constructors

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

drisspg commented on torch/nested/__init__.py:
```diff
@@ -1,11 +1,89 @@
+        raise TypeError("nested_tensor(): Expected first argument to be a list of tensors ")
+    new_data = []
+    for t in tensor_list:
+        new_data.append(t.clone().detach())
```
Small performance hit by having to clone but I think it is worth it in order to do this all in python

cpuhrsch replied:
Why do we have to clone?

mikaylagawarecki replied:
Should it just be `t.detach()`?

cpuhrsch replied:
Yeah I think that's enough

mikaylagawarecki replied:
Ah I see will change, I understand that `cat` in NestedTensorUtils:wrap_tensor_node will ultimately create a copy, but in practice I've never seen `detach` used without `clone`

drisspg replied:
In the readme: 
```
# NOTE: Previously, if we change the tensor metadata (e.g. sizes / strides /
# storage / storage_offset) of a tensor created from `detach()`, those metadata
# in the original tensor will also be updated. However, the new behavior is that
# those metadata changes to the detached tensor will not update the original tensor
# anymore, and in the `detach()` function we need to set `allow_tensor_metadata_change_`
# to false to make such changes explicitly illegal, in order to prevent users from
# changing metadata of the detached tensor and expecting the original tensor to also
# be updated.
```
I think the concern is that we don't want the detach to mutate the original tensor but like you said since wrap_tensor makes a copy I think in this case it is okay.

mikaylagawarecki replied:
Ah gotcha thanks Driss :)

drisspg reviewed (approved):
Didn't look at all the call sites of `torch.nested.nested..` but looks good. @cpuhrsch do you have issue with the two functions created at the python level calling into the same native function

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

mehtanirav commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`"

Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.



Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

[ghstack-poisoned]
